src/Numerics/Mesh/DSS.jl

```diff
@@ -0,0 +1,318 @@
+    #----Direct Stiffness Summation-------------
+    vertmap = grid.vertmap
+    edgemap = grid.edgemap
+    facemap = grid.facemap
+
+    vtconn = grid.topology.vtconn
+    fcconn = grid.topology.fcconn
+    edgconn = grid.topology.edgconn
+
+    vtconnoff = grid.topology.vtconnoff
+    edgconnoff = grid.topology.edgconnoff
+
+    nvt, nfc, nedg =
+        length(vtconnoff) - 1, size(fcconn, 1), length(edgconnoff) - 1
+
+    Nq = polynomialorders(grid) .+ 1
+    Nqmax = maximum(Nq)
+    Nemax = Nqmax - 2
+    Nfmax = size(facemap, 1)
+    args = (
+        Q.data,
+        vertmap,
+        edgemap,
+        facemap,
+        vtconn,
+        vtconnoff,
+        edgconn,
+        edgconnoff,
+        fcconn,
+        nvt,
+        nedg,
+        nfc,
+        Nemax,
+        Nfmax,
+    )
+    if device == CPU()
+        dss3d_CPU!(args...)
+    else
+        n_items = nvt + nfc + nedg
+        tx = max(n_items, max_threads)
+        bx = cld(n_items, tx)
+        @cuda threads = (tx) blocks = (bx) dss3d_CUDA!(args...)
+    end
+    return nothing
+end
+
+function dss3d_CUDA!(
+    data,
+    vertmap,
+    edgemap,
+    facemap,
+    vtconn,
+    vtconnoff,
+    edgconn,
+    edgconnoff,
+    fcconn,
+    nvt,
+    nedg,
+    nfc,
+    Nemax,
+    Nfmax,
+)
+    I = eltype(nvt)
+    FT = eltype(data)
+
+    tx = threadIdx().x        # threaid id
+    bx = blockIdx().x         # block id
+    bxdim = blockDim().x      # block dimension
+    glx = tx + (bx - 1) * bxdim # global id
+    nvars = size(data, 2)
+
+    # A mesh node is either
+    # - interior (no dss required)
+    # - on an element corner / vertex,
+    # - edge (excluding end point element corners / vertices)
+    # - face (excluding edges and corners)
+    if glx ≤ nvt #vertex DSS
+        vx = glx
+        for ivar in 1:nvars
+            dss_vertex!(vtconn, vtconnoff, vertmap, data, ivar, vx, FT)
+        end
+    elseif glx > nvt && glx ≤ (nvt + nedg) # edge DSS
+        ex = glx - nvt
+        for ivar in 1:nvars
+            dss_edge!(edgconn, edgconnoff, edgemap, Nemax, data, ivar, ex, FT)
+        end
+    elseif glx > (nvt + nedg) && glx ≤ (nvt + nedg + nfc) # face DSS
+        fx = glx - (nvt + nedg)
+        for ivar in 1:nvars
+            dss_face!(fcconn, facemap, Nfmax, data, ivar, fx)
+        end
+    end
+
+    return nothing
+end
+
+function dss3d_CPU!(
+    data,
+    vertmap,
+    edgemap,
+    facemap,
+    vtconn,
+    vtconnoff,
+    edgconn,
+    edgconnoff,
+    fcconn,
+    nvt,
+    nedg,
+    nfc,
+    Nemax,
+    Nfmax,
+)
+    I = eltype(nvt)
+    FT = eltype(data)
+    nvars = size(data, 2)
+
+    # A mesh node is either
+    # - interior (no dss required)
+    # - on an element corner / vertex,
+    # - edge (excluding end point element corners / vertices)
+    # - face (excluding edges and corners)
+    for ivar in 1:nvars
+        for vx in 1:nvt # vertex DSS
+            dss_vertex!(vtconn, vtconnoff, vertmap, data, ivar, vx, FT)
```

Comment:
how does this handle boundary faces?

Reply:
Boundary faces are skipped as dss is not needed.  Same for vertices lying on a single element.